File: solutions/csharp/zipper/1/Zipper.cs
using System;

// ReSharper disable once CheckNamespace
public class BinTree
{
	public BinTree(int value, BinTree? left, BinTree? right)
	{
		Value = value;
		Left = left;
		Right = right;
	}

	public int Value { get; }
	public BinTree? Left { get; }
	public BinTree? Right { get; }

	public override bool Equals(object? obj)
		=> obj is BinTree other
		   && (Value == other.Value)
		   && Equals(Left, other.Left)
		   && Equals(Right, other.Right);

	public override int GetHashCode()
		=> HashCode.Combine(Value, Left, Right);
}

// ReSharper disable once CheckNamespace
public sealed class Zipper
{
	private readonly BinTree _focus;
	private readonly Zipper? _parent;
	private readonly bool _isLeftChild;

	private Zipper(BinTree focus, Zipper parent, bool isLeftChild)
	{
		_focus = focus;
		_parent = parent;
		_isLeftChild = isLeftChild;
	}

	public int Value()
		=> _focus.Value;

	public Zipper SetValue(int newValue)
	{
		BinTree newFocus = new(newValue, _focus.Left, _focus.Right);
		return new(newFocus, UpdateParent(newFocus)!, _isLeftChild);
	}

	public Zipper SetLeft(BinTree binTree)
	{
		BinTree newFocus = new(_focus.Value, binTree, _focus.Right);
		return new(newFocus, UpdateParent(newFocus)!, _isLeftChild);
	}

	public Zipper SetRight(BinTree binTree)
	{
		BinTree newFocus = new(_focus.Value, _focus.Left, binTree);
		return new(newFocus, UpdateParent(newFocus)!, _isLeftChild);
	}

	public Zipper? Left()
		=> _focus.Left == null
			? null
			: new Zipper(_focus.Left, this, true);

	public Zipper? Right()
		=> _focus.Right == null
			? null
			: new Zipper(_focus.Right, this, false);

	public Zipper Up()
		=> _parent!;

	public BinTree ToTree()
		=> _parent == null
			? _focus
			: _parent.ToTree();

	public static Zipper FromTree(BinTree tree)
		=> new(tree, null!, false);

	private Zipper? UpdateParent(BinTree newChild)
	{
		if (_parent == null)
		{
			return null;
		}

		BinTree newParentTree = _isLeftChild
			? new(_parent._focus.Value, newChild, _parent._focus.Right)
			: new BinTree(_parent._focus.Value, _parent._focus.Left, newChild);

		Zipper newParentZipper = new(newParentTree, _parent._parent!, _parent._isLeftChild);

		return _parent._parent == null
			? newParentZipper
			: newParentZipper.UpdateParent(newParentTree);
	}

	public override bool Equals(object? obj)
		=> obj is Zipper other
		   && Equals(_focus, other._focus)
		   && Equals(_parent, other._parent)
		   && (_isLeftChild == other._isLeftChild);

	public override int GetHashCode()
		=> HashCode.Combine(_focus, _parent, _isLeftChild);
}

File: solutions/scala/zipper/1/src/main/scala/Zipper.scala
import scala.annotation.tailrec

case class BinTree[A](value: A, left: Option[BinTree[A]], right: Option[BinTree[A]])

final case class ZipperState[A](
                                   focus: BinTree[A],
                                   parent: Option[ZipperState[A]],
                                   isLeftChild: Boolean
                               ) {

    def moveFocus(newFocus: BinTree[A]): ZipperState[A] =
        val updatedParent = updateParent(newFocus, parent, isLeftChild)
        ZipperState(newFocus, updatedParent, isLeftChild)

    @tailrec
    private def updateParent(
                                   newChild: BinTree[A],
                                   parent: Option[ZipperState[A]],
                                   isLeftChild: Boolean
                               ): Option[ZipperState[A]] = {

        parent match {
            case None => None

            case Some(pattern) =>
                val newParentFocus =
                    if (isLeftChild)
                        BinTree(pattern.focus.value, Some(newChild), pattern.focus.right)
                    else
                        BinTree(pattern.focus.value, pattern.focus.left, Some(newChild))

                val rebuiltParent =
                    ZipperState(
                        newParentFocus,
                        pattern.parent,
                        pattern.isLeftChild
                    )

                pattern.parent match
                    case None => Some(rebuiltParent)
                    case Some(_) =>
                        updateParent(newParentFocus, pattern.parent, pattern.isLeftChild)
        }
    }
}

object Zipper:
    def fromTree[A](bt: BinTree[A]): ZipperState[A] =
        ZipperState(bt, None, isLeftChild = false)
    def value[A](zipper: ZipperState[A]): A =
        zipper.focus.value

    def setValue[A](value: A, zipper: ZipperState[A]): ZipperState[A] = {
        val newFocus = BinTree(value, zipper.focus.left, zipper.focus.right)
        zipper.moveFocus(newFocus)
    }

    def setLeft[A](left: Option[BinTree[A]], zipper: ZipperState[A]): ZipperState[A] = {
        val newFocus = BinTree(zipper.focus.value, left, zipper.focus.right)
        zipper.moveFocus(newFocus)
    }

    def setRight[A](right: Option[BinTree[A]], zipper: ZipperState[A]): ZipperState[A] = {
        val newFocus = BinTree(zipper.focus.value, zipper.focus.left, right)
        zipper.moveFocus(newFocus)
    }

    private def move[A](
                           zipper: ZipperState[A],
                           nextFocus: ZipperState[A] => Option[BinTree[A]],
                           isLeftChild: Boolean
                       ): Option[ZipperState[A]] =
        nextFocus(zipper).map { child =>
            ZipperState(child, Some(zipper), isLeftChild)
        }

    def left[A](zipper: ZipperState[A]): Option[ZipperState[A]] =
        move(zipper, _.focus.left, isLeftChild = true)

    def right[A](zipper: ZipperState[A]): Option[ZipperState[A]] =
        move(zipper, _.focus.right, isLeftChild = false)

    def up[A](zipper: ZipperState[A]): Option[ZipperState[A]] =
        zipper.parent

    @tailrec
    def toTree[A](zipper: ZipperState[A]): BinTree[A] =
        zipper.parent match
            case None => zipper.focus
            case Some(p) => toTree(p)


